File: sparkcql-core/src/main/scala/gov/va/sparkcql/adapter/library/FileLibraryAdapter.scala
package gov.va.sparkcql.adapter.library

import scala.reflect.runtime.universe._
import gov.va.sparkcql.io.Files
import org.json4s._
import org.json4s.jackson.Serialization.{read, write}
import javax.xml.namespace.QName
import org.apache.spark.sql.{SparkSession, DataFrame}
import gov.va.sparkcql.compiler.CompilerGateway
import gov.va.sparkcql.types._
import gov.va.sparkcql.logging.Log
import org.hl7.elm.r1.VersionedIdentifier

class FileLibraryAdapter(path: String) extends LibraryAdapter {

  assert(path != null)
  assert(path != "")

  val currentDir = Files.currentDir()
  
  lazy val contents = {
    Files.search(path, ".cql").flatMap(file => {
      val content = scala.io.Source.fromFile(file).mkString
      val identifier = CompilerGateway.parseVersionedIdentifier(content)
      Map(identifier -> content)
    }).toMap
  }

  def getLibraryContent(identifier: VersionedIdentifier): Option[String] = contents.get(identifier)
}

File: sparkcql-core/src/main/java/gov/va/sparkcql/adapter/library/FileLibraryAdapter.java
package gov.va.sparkcql.adapter.library;

import java.util.Map;
import java.util.stream.Collectors;

import org.hl7.elm.r1.VersionedIdentifier;

import gov.va.sparkcql.compiler.CompilerGateway;

import gov.va.sparkcql.common.Directory;

public class FileLibraryAdapter implements LibraryAdapter {

    protected Map<VersionedIdentifier, String> contents;

    public FileLibraryAdapter(String path) {
        if (path == null || path == "") {
            contents = Map.of();
        } else {
            var compilerGateway = new CompilerGateway();
            contents = Directory
                .find(path, ".cql")
                .map(Directory::readString) 
                .collect(Collectors.toMap(compilerGateway::parseVersionedIdentifier, c -> c));
        }
    }

    @Override
    public String getLibrarySource(VersionedIdentifier identifier) {
        if (!contents.containsKey(identifier)) {
            throw new RuntimeException("Library " + identifier.toString() + " not found.");
        }
        return contents.get(identifier);
    }

    @Override
    public Boolean containsLibrary(VersionedIdentifier identifier) {
        return contents.containsKey(identifier);
    }
}